Executing `--dry-run` via shell `$()` can fail when certain JVM options are used.

**Jetty version(s)**
12.0.23

**Jetty Environment**
ee8

**Java version/vendor** `(use: java -version)`
openjdk version "17.0.12" 2024-07-16
OpenJDK Runtime Environment Temurin-17.0.12+7 (build 17.0.12+7)
OpenJDK 64-Bit Server VM Temurin-17.0.12+7 (build 17.0.12+7, mixed mode, sharing)

**OS type/version**
Ubuntu 24.04.1 LTS

**Description**
I am unable to change the values for the different options for the G1 garbage collector such as -XX:MaxGCPauseMillis=200

However, I am able to enable options such as -XX:+UseG1GC and -XX:+UseStringDeduplication and -XX:+PrintGCDetails but unable to change values

It gives me the following error:

<img width="1511" height="576" alt="Image" src="https://github.com/user-attachments/assets/66bf06fa-eff7-403c-9ea3-a5c1dc4452ef" />

**How to reproduce?**
1) Define a custom Jetty module (see https://jetty.org/docs/jetty/12/operations-guide/start/index.html#configure-custom-module-exec) and specify the following JVM options:

-XX:+UseG1GC
-XX:MaxGCPauseMillis=200

2) Copy the JVM module to $JETTY_BASE/modules

3) Enable it using 'java -jar $JETTY_HOME/start.jar --add-modules=jvm'

4) Start Jetty using '$(java -jar $JETTY_HOME/start.jar --dry-run)' to fork only one JVM

I can replicate.

I didn't specify a custom jvm module tho.

```
[bases]$ mkdir jvm-12
[jvm-12]$ java -jar ../../jetty-home-12.0.23/start.jar --add-modules=http,jvm
INFO  : mkdir ${jetty.base}/start.d
INFO  : jvm             initialized in ${jetty.base}/start.d/jvm.ini
INFO  : server          transitively enabled, ini template available with --add-modules=server
INFO  : logging-jetty   transitively enabled
INFO  : http            initialized in ${jetty.base}/start.d/http.ini
INFO  : resources       transitively enabled
INFO  : threadpool      transitively enabled, ini template available with --add-modules=threadpool
INFO  : logging/slf4j   dynamic dependency of logging-jetty
INFO  : bytebufferpool  transitively enabled, ini template available with --add-modules=bytebufferpool
INFO  : mkdir ${jetty.base}/resources
INFO  : copy ${jetty.home}/modules/logging/jetty/resources/jetty-logging.properties to ${jetty.base}/resources/jetty-logging.properties
INFO  : Base directory was modified

[jvm-12]$ gvim start.d/jvm.ini 
[jvm-12]$ grep -v "#" start.d/jvm.ini 
--modules=jvm

--exec
-XX:+UseG1GC
-XX:MaxGCPauseMillis=200

[jvm-12]$ java --version
openjdk 17.0.15 2025-04-15
OpenJDK Runtime Environment Temurin-17.0.15+6 (build 17.0.15+6)
OpenJDK 64-Bit Server VM Temurin-17.0.15+6 (build 17.0.15+6, mixed mode, sharing)

[jvm-12]$ java -jar ../../jetty-home-12.0.23/start.jar
WARN  : Forking second JVM due to forking module(s): []. Use --dry-run to generate the command line to avoid forking.
2025-07-16 08:51:21.370:INFO :oejs.Server:main: jetty-12.0.23; built: 2025-07-02T14:02:02.445Z; git: 01a4119797e9cee53c974ae126cc316d0c8a533a; jvm 17.0.15+6
2025-07-16 08:51:21.407:INFO :oejs.AbstractConnector:main: Started ServerConnector@2fd6b6c7{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
2025-07-16 08:51:21.421:INFO :oejs.Server:main: Started oejs.Server@4d826d77{STARTING}[12.0.23,sto=0] @512ms
<Ctrl+C>

[jvm-12]$ $(java -jar $JETTY_HOME/start.jar --dry-run)
Error: Could not find or load main class '-XX:MaxGCPauseMillis=200'
Caused by: java.lang.ClassNotFoundException: '-XX:MaxGCPauseMillis=200'

[jvm-12]$ java -jar $JETTY_HOME/start.jar --dry-run
/home/joakim/java/jvm/jdk-17.0.15+6/bin/java -Djava.io.tmpdir=/tmp -Djetty.home=/home/joakim/code/jetty/distros/jetty-home-12.0.23 -Djetty.base=/home/joakim/code/jetty/distros/bases/jvm-12 -XX:+UseG1GC '-XX:MaxGCPauseMillis=200' --class-path /home/joakim/code/jetty/distros/bases/jvm-12/resources:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/logging/slf4j-api-2.0.17.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/logging/jetty-slf4j-impl-12.0.23.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/jetty-http-12.0.23.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/jetty-server-12.0.23.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/jetty-xml-12.0.23.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/jetty-util-12.0.23.jar:/home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/jetty-io-12.0.23.jar org.eclipse.jetty.xml.XmlConfiguration java.version=17.0.15 jetty.base=/home/joakim/code/jetty/distros/bases/jvm-12 jetty.base.uri=file:///home/joakim/code/jetty/distros/bases/jvm-12 jetty.home=/home/joakim/code/jetty/distros/jetty-home-12.0.23 jetty.home.uri=file:///home/joakim/code/jetty/distros/jetty-home-12.0.23 jetty.webapp.addHiddenClasses=org.eclipse.jetty.logging.,file:///home/joakim/code/jetty/distros/jetty-home-12.0.23/lib/logging/,org.slf4j. runtime.feature.alpn=true slf4j.version=2.0.17 /home/joakim/code/jetty/distros/jetty-home-12.0.23/etc/jetty-bytebufferpool.xml /home/joakim/code/jetty/distros/jetty-home-12.0.23/etc/jetty-threadpool.xml /home/joakim/code/jetty/distros/jetty-home-12.0.23/etc/jetty.xml /home/joakim/code/jetty/distros/jetty-home-12.0.23/etc/jetty-http.xml
```

The quoting used in the argument ...

```
'-XX:MaxGCPauseMillis=200'
```

Is being interpreted by the shell not as a java option when using the `$()` technique., but something else.
The JVM is starting up thinking that this entry is a `<mainclass>` instead.

This can be shown in this simple demo...

```
[tmp]$ mkdir shell-test
[tmp]$ cd shell-test
[shell-test]$ gvim Main.java
[shell-test]$ cat Main.java 
public class Main {
    public static void main(String args[]) {
        System.out.println("Main Executing");
    }
}
[shell-test]$ javac Main.java
[shell-test]$ ls -la
total 16
drwxrwxr-x 2 joakim joakim 4096 Jul 16 09:08 .
drwxr-xr-x 6 joakim joakim 4096 Jul 16 09:07 ..
-rw-rw-r-- 1 joakim joakim  416 Jul 16 09:08 Main.class
-rw-rw-r-- 1 joakim joakim  119 Jul 16 09:08 Main.java

[shell-test]$ java -cp . Main
Main Executing

[shell-test]$ java -XX:+UseG1GC '-XX:MaxGCPauseMillis=200' -cp . Main
Main Executing

[shell-test]$ gvim cmdline.txt
[shell-test]$ cat cmdline.txt 
java -XX:+UseG1GC '-XX:MaxGCPauseMillis=200' -cp . Main
[shell-test]$ $(cat cmdline.txt)
Error: Could not find or load main class '-XX:MaxGCPauseMillis=200'
Caused by: java.lang.ClassNotFoundException: '-XX:MaxGCPauseMillis=200'
```

It doesn't matter the quoting style, on bash, all 3 quoting styles fail this use case.

``` shell
# Using tick (default in Jetty)
[shell-test]$ cat cmdline.txt 
java -XX:+UseG1GC '-XX:MaxGCPauseMillis=200' -cp . Main
[shell-test]$ $(cat cmdline.txt)
Error: Could not find or load main class '-XX:MaxGCPauseMillis=200'
Caused by: java.lang.ClassNotFoundException: '-XX:MaxGCPauseMillis=200'

# Using double quote
[shell-test]$ cat cmdline.txt 
java -XX:+UseG1GC "-XX:MaxGCPauseMillis=200" -cp . Main
[shell-test]$ $(cat cmdline.txt)
Error: Could not find or load main class "-XX:MaxGCPauseMillis=200"
Caused by: java.lang.ClassNotFoundException: "-XX:MaxGCPauseMillis=200"

# Using single quote
[shell-test]$ cat cmdline.txt 
java -XX:+UseG1GC `-XX:MaxGCPauseMillis=200` -cp . Main
[shell-test]$ $(cat cmdline.txt)
Error: Could not find or load main class `-XX:MaxGCPauseMillis=200`
Caused by: java.lang.ClassNotFoundException: `-XX:MaxGCPauseMillis=200`
```

--- a/documentation/jetty/modules/operations-guide/pages/start/index.adoc
+++ b/documentation/jetty/modules/operations-guide/pages/start/index.adoc
@@ -305,14 +305,7 @@ args=--dry-run
 replace=( ),$1\\\n
 ----
 
-You can then run the generated command line.
-
-For example, in the Linux `bash` shell you can run it by wrapping it into `$(\...)`:
-
-[source,bash]
-----
-$ $(java -jar $JETTY_HOME/start.jar --dry-run)
-----
+You can then run the generated command line by copying the output generated using the `--dry-run` option, and pasting it in a terminal window.
 
 The `--dry-run` option is quite flexible and below you can find a few examples of how to use it to avoid forking a second JVM, or generating scripts or creating an arguments file that can be passed to (a possibly alternative) `java` executable.
 
@@ -427,29 +420,50 @@ replace=( ),$1\\\n
 
 Note how the program arguments are a list of properties in the form `<name>=<value>` and a list of Jetty XML files.
 
-The various parts of the full JVM command line can be combined to leverage the arguments file feature (that is, specify the JVM options in a file rather than on the command line) that is built-in in the `java` executable:
+To display the xref:deploy/index.adoc[environments] configuration:
 
 [source,bash,subs=+quotes]
 ----
-$ java -jar $JETTY_HOME/start.jar --dry-run=##opts,path,main,args## > /tmp/jvm_cmd_line.txt
+$ java -jar $JETTY_HOME/start.jar --dry-run=##envs##
+----
+
+[jetty%nowrap]
+----
+[jetty]
+setupModules=code:example$jetty-modules/postgresql.mod
+setupArgs=--add-modules=http,ee10-deploy
+args=--dry-run=args
+replace=( ),$1\\\n
+----
+
+Note how the environments configuration specifies additional class-path and Jetty XML files for every environment.
+
+The various parts of the full JVM command line can be combined to leverage the JVM arguments file feature (that is, specify the JVM options in a file rather than on the command line) that is built-in in the `java` executable:
+
+[source,bash,subs=+quotes]
+----
+$ java -jar $JETTY_HOME/start.jar --dry-run=##opts,path,main,args,envs## > /tmp/jvm_cmd_line.txt
 $ /some/other/java @/tmp/jvm_cmd_line.txt
 ----
 
-Using `--dry-run=opts,path,main,args` can be used to avoid that the Jetty start mechanism forks a second JVM when using modules that require forking:
+Using `--dry-run=opts,path,main,args,envs` can be used to avoid that the Jetty start mechanism forks a second JVM when using modules that require forking.
+
+In the Linux `bash` shell you can run the output of `--dry-run` in this way:
 
 [source,bash]
 ----
-$ java $(java -jar $JETTY_HOME/start.jar --dry-run=opts,path,main,args)
+$ java -jar $JETTY_HOME/start.jar --dry-run=opts,path,main,args,envs | xargs java
 ----
 
-The output of different `--dry-run` executions can be creatively combined in a shell script:
+The output of different `--dry-run` executions can be creatively combined in a script, for example in the Linux `bash` shell:
 
 [source,bash,subs=+quotes]
 ----
 $ OPTS=$(java -jar start.jar --dry-run=##opts,path##)
 $ MAIN=$(java -jar start.jar --dry-run=##main##)
 $ ARGS=$(java -jar start.jar --dry-run=##args##)
-$ java $OPTS -Dextra=opt $MAIN $ARGS extraProp=value extra.xml
+$ ENVS=$(java -jar start.jar --dry-run=##envs##)
+$ echo $OPTS -Dextra=opt $MAIN $ARGS $ENVS extraProp=value extra.xml | xargs java
 ----
 
 [[start]]

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -567,7 +567,7 @@ then
   fi
 fi
 
-# Collect the dry-run (of opts,path,main,args) from the jetty.base configuration
+# Collect the dry-run (of opts,path,main,args,envs) from the jetty.base configuration
 JETTY_DRY_RUN=$(echo "${JETTY_ARGS[*]} ${JAVA_OPTIONS[*]}" | xargs "$JAVA" -jar "$JETTY_START" --dry-run=opts,path,main,args,envs)
 RUN_ARGS=($JETTY_SYS_PROPS ${JETTY_DRY_RUN[@]})
 

